homeassistant/components/switchbee/switch.py

```diff
@@ -0,0 +1,141 @@
+"""Support for SwitchBee switch."""
+import logging
+from typing import Any
+
+from switchbee import SWITCHBEE_BRAND
+from switchbee.api import SwitchBeeDeviceOfflineError, SwitchBeeError
+from switchbee.device import ApiStateCommand, DeviceType, SwitchBeeBaseDevice
+
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import SwitchBeeCoordinator
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Switchbee switch."""
+    coordinator: SwitchBeeCoordinator = hass.data[DOMAIN][entry.entry_id]
+    device_types = (
+        [DeviceType.TimedPowerSwitch]
+        if coordinator.switch_as_light
+        else [
+            DeviceType.TimedPowerSwitch,
+            DeviceType.GroupSwitch,
+            DeviceType.Switch,
+            DeviceType.TimedSwitch,
+            DeviceType.TwoWay,
+        ]
+    )
+
+    async_add_entities(
+        Device(hass, device, coordinator)
+        for device in coordinator.data.values()
+        if device.type in device_types
+    )
+
+
+class Device(CoordinatorEntity, SwitchEntity):
+    """Representation of an Switchbee switch."""
+
+    def __init__(self, hass, device: SwitchBeeBaseDevice, coordinator):
+        """Initialize the Switchbee switch."""
+        super().__init__(coordinator)
+        self._session = aiohttp_client.async_get_clientsession(hass)
+        self._attr_name = f"{device.name}"
+        self._device_id = device.id
+        self._attr_unique_id = f"{coordinator.mac_formated}-{device.id}"
+        self._attr_is_on = False
+        self._attr_available = True
+        self._attr_has_entity_name = True
+        self._device = device
+        self._attr_device_info = DeviceInfo(
+            name=f"SwitchBee_{str(device.unit_id)}",
+            identifiers={
+                (
+                    DOMAIN,
+                    f"{str(device.unit_id)}-{coordinator.mac_formated}",
+                )
+            },
+            manufacturer=SWITCHBEE_BRAND,
+            model=coordinator.api.module_display(device.unit_id),
+            suggested_area=device.zone,
+            via_device=(
+                DOMAIN,
+                f"{coordinator.api.name} ({coordinator.api.mac})",
+            ),
+        )
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+
+        async def async_refresh_state():
+
+            try:
+                await self.coordinator.api.set_state(self._device_id, "dummy")
```

Comment:
The coordinator fetches the data from the Central Unit and throw it as is to the entity, the entity responsible for taking care of the new status.

I don't see how I can migrate this code out of the entity to the coordinator... I'd be happy for example

Reply:
If the device lost state is not specific to the switch platform, the coordinator should probably check for this case for all devices and try to set the dummy state before returning the device state and notifying the entities. The entities can just check their state and set unavailable if needed.

Reply:
Setting the dummy state will be relevant for the next polling.

If the coordinator start refreshing the state of the offline devices, wouldn't this delay the state update for the platforms?

Bottom line I think you are right and this is more Coordinator work than platform and I will address it.

The question if can I address it in a PR post https://github.com/home-assistant/core/pull/78412?

I think I have already overloaded https://github.com/home-assistant/core/pull/78412 with changes

Reply:
It's ok to do in a separate PR. 👍

Some thoughts: Does the set state api call take a long time to complete? How common is this problem? Can we gather the calls and await them concurrently? Creating a task for the api calls will allow the entities to update state faster but may complicate the coordinator if we need to manage these tasks to not have them lingering before the next task is created. We can discuss more in the PR when it's up.

Reply:
Few milliseconds, this happens in case of the Central Unit (Hub) lost rf communication with the wall outlet (unlikely to happen in a steady state) or of the outlet power was lost (also unlikely)

I think it's possible to gather up all the devices that lost communication andcrun task concurrently but its to much complications for a scenario that unlikely to happen.

Sure, I'll make sure to mention you so we come up with efficient solution.

Thanks a lot!